### DISCUSSION TITLE
Web Export

### DISCUSSION THREAD

**Discussion by abueide** [libktx/ktx#305] [Q&A] [answered]
Are there any viable solutions to export to web? Or is all hope lost until korge is good enough? Even if it was though, you have to sacrifice all the cool java libraries built over the years for kotlin multiplatform. I've heard rumours about teavm or something called dragome but I don't know much about them.

**Comment by czyzby** [comment, marked as answer]
- [TeaVM](https://github.com/konsoletyper/teavm-libgdx) is an alternative to GWT, but I think the LibGDX backend is just a proof of concept rather than a working solution. You can go through the forks to see if anyone picked up after the TeaVM author. As a side note, TeaVM itself is a really cool project, probably with the highest quality among similar libraries.
- [JTransc](https://github.com/jtransc/gdx-backend-jtransc) transpiles JVM bytecode to Haxe and then to Flash, JavaScript and a couple other platforms. Again, I'm not sure how feature-complete this is, but it's worth a try. I think the author moved on to create Korge, so there's that.
- [xpenatan](https://github.com/xpenatan/gdx-web-tools) maintains a few web backends, with Dragome backend being the most feature-complete (I think). It might be your best bet as far as working solutions go, but I believe the code quality is just not the best. For a while it used to be an Eclipse project without Maven/Gradle setup, mostly copying and modifying existing GWT sources without tests, refactoring, etc. Dragome never took off and I don't think it's actively maintained anymore.

---

One thing to note though is that web games written in other languages than JS/TS tend to be harder to debug and more error-prone. It's cool that you can write your game once and export for multiple platforms, but realistically - even with pure Java and no complex LibGDX extensions - supporting the web is a pain. I don't think I've ever had a worse experience debugging my code than with GWT.

LibGDX and KTX are the best fit for mobile games. If you want to focus on web games, I'm afraid JS frameworks might be your best bet.

**Reply by czyzby** [reply]
See #443 for an example of a KTX game exported to HTML/WebGL with the xpenatan's backend.

**Comment by abueide** [comment]
Thanks for the detailed overview. I just recently learned that web games are stuck to being single threaded even if I did find a solution, which is a necessity for my game. It looks like I just have to give up on supporting web, which is kind of a relief tbh, I was not looking forward to all the issues that come with supporting web.

**Comment by czyzby** [comment]
Well, there are web workers that you could use, but since they cannot share state like Java threads, you'd probably have to rewrite significant portions of your game. You can also simulate "green threads" or coroutines with async operations, similar to how KotlinJS/TeaVM achieve this despite JS being single-threaded, but obviously that isn't true parallel processing.

> It looks like I just have to give up on supporting web, which is kind of a relief tbh, I was not looking forward to all the issues that come with supporting web.

In my experience, if you want to support the web, you should start with the web, as it's the most limited platform by far. Even with plain Java + LibGDX you need to constantly track GWT compilation and performance if you don't want to end up with an unplayable application.

**Comment by CoenRijsdijk** [comment]
I'm currently working on this, using Bytecoder and WebGl. I'm working on it until 30 January 2021(my traineeship for Computer Science Bsc degree). Check my project and latest findings here: https://github.com/squins/libgdx-wasm-with-bytecoder
We really would like to have a LibGDX backend for web that can be used with Kotlin. That is the goal :smile:.
I am now working on getting audio and WebGl to work.

**Comment by czyzby** [comment]
@CoenRijsdijk That's a very cool project, keep us updated!
